cirq-core/cirq/sim/act_on_args.py

```diff
@@ -139,6 +140,36 @@ def log_of_measurement_results(self) -> Dict[str, Any]:
     def qubits(self) -> Tuple['cirq.Qid', ...]:
         return self._qubits
 
+    def swap(self, q1: 'cirq.Qid', q2: 'cirq.Qid', *, inplace=False):
+        """Swaps two qubits.
+
+        This only affects the index, and does not modify the underlying
+        state."""
+        if q1.dimension != q2.dimension:
+            raise ValueError(f'Cannot swap different dimensions: q1={q1}, q2={q2}')
+
+        args = self if inplace else copy.copy(self)
```

Comment:
@95-martin-orion Are you okay with `copy.copy` here for a shallow copy? I don't see it used much elsewhere. Another option would be to make ActOnArgs and descendants all dataclasses and use `dataclasses.replace` here. I think it makes sense to do so if we're already using dataclasses elsewhere, though I don't have any experience with dataclasses so I might be wrong. (For instance, there's no sense in making them frozen, because the state they represent changes and would be too expensive to copy for each change, but it doesn't look like being frozen is a requirement for a dataclass).

Note a fair chunk of #4265 is also "cleaning stuff up" by using `copy.copy`, and should probably be split into two separate PRs. I just wanted to reduce the amount duplication I noticed while adding CH-form.

Reply:
`copy.copy` is fine for shallow copy, but is that what we want here? Sharing a `prng` reference between two `ActOnArgs` could generate non-reproducible results.

We have a few other `dataclass`es in Cirq, so I'd be fine with using it for `ActOnArgs` - although in my experience `dataclass` inheritance with default values can be awkward.

Reply:
It behaves like methods previously known as split, join, and reindex, which also do shallow copies (though those methods create the copy manually rather than use copy.copy -- changing those to use `copy.copy` is the "cleanup" section of #4265 I mentioned above). The prng is "owned" by the simulator itself, so they all keep a reference to that.

Reply:
I guess there's an implicit assumption here that the calling `ActOnArgs` object will be discarded afterwards? I think the `inplace` argument is what's confusing me - if both `inplace=True` and `inplace=False` will invalidate the old object, why should we have an `inplace=False` mode at all?

Reply:
As an aside: if the implicit assumption above is correct, could you update the docstrings of both this and the split/join/reindex methods to clarify this?

Reply:
Allowing swap to be inplace cleans up the code a lot in this PR, because otherwise we'd have to go through and backfill all the dictionaries afterward with the new reference in several places. But the existing methods (split/join/reorder) all create new copies instead of operating inplace. It seemed awkward to introduce a new semantic just for this method.

So rather than having different semantics for different methods, or to have a single semantic that may be suboptimal for some cases, I figured the best approach would be to standardize on a parameter for all these methods that lets the user choose, especially since it only requires a single line of code. (I added this parameter to the remaining functions in #4265). This is also the approach numpy itself uses in a lot of functions.

Yes I think it's good to make that explicit. I'll update the docstrings, and then reupdate them in #4265.